src/hooks/usePortal.ts

```diff
@@ -0,0 +1,303 @@
+import { useSendTransaction, useWalletClient } from "wagmi";
+import { useChain } from "./useChain";
+import {
+  UseMutateAsyncFunction,
+  useMutation,
+  useQuery,
+} from "@tanstack/react-query";
+import { arbitrum, fantom, mainnet, optimism } from "viem/chains";
+import { parseUnits } from "viem";
+import { ZERO_ADDRESS } from "@/lib/constants";
+
+type PermitTypes = {
+  Permit: [
+    { name: "owner"; type: "address" },
+    { name: "spender"; type: "address" },
+    { name: "value"; type: "uint256" },
+    { name: "nonce"; type: "uint256" },
+    { name: "deadline"; type: "uint256" },
+  ];
+};
+
+type PermitTx = {
+  domain: {
+    name: string;
+    version: string;
+    chainId: number;
+    verifyingContract: `0x${string}`;
+  };
+  types: PermitTypes;
+  value: {
+    owner: `0x${string}`;
+    spender: `0x${string}`;
+    value: bigint;
+    nonce: bigint;
+    deadline: bigint;
+  };
+};
+
+type TransactionToSubmit = {
+  to: `0x${string}`;
+  from: `0x${string}`;
+  data: `0x${string}`;
+  gasLimit: string;
+  value?: bigint;
+};
+
+type PortalQuoteParams = {
+  approveInputToken: UseMutateAsyncFunction;
+  shouldApprove: unknown;
+  gaslessSignature: `0x${string}` | undefined;
+  inputToken: string;
+  inputTokenDecimals: number;
+  inputAmount: string;
+  outputToken: string;
+  sender: string;
+};
+
+export type PortalQuoteResponse = {
+  tx: TransactionToSubmit;
+  context: PortalQuoteContext;
+};
+
+export type PortalQuoteContext = {
+  orderId: string;
+  minOutputAmount: string;
+  minOutputAmountUsd: number;
+  slippageTolerancePercentage: number;
+  gasLimit: string;
+  inputAmount: string;
+  inputAmountUsd: number;
+  inputToken: string;
+  outputToken: string;
+  outputAmount: string;
+  outputAmountUsd: number;
+  partner: string;
+  feeToken: string;
+  feeAmount: string;
+  feeAmountUsd: number;
+  sender: `0x${string}`;
+  recipient: `0x${string}`;
+  target: `0x${string}`;
+  value: string;
+  route: string[];
+  steps: string[];
+};
+
+export type PortalTransactionResult =
+  | { hash: string; signature?: undefined }
+  | { signature: string; hash?: undefined };
+
+const PORTAL_API_BASE_URI = "https://api.portals.fi/v2";
+const PORTALS_CHAIN_ID = {
+  [mainnet.id]: "ethereum",
+  [arbitrum.id]: "arbitrum",
+  [optimism.id]: "optimism",
+  [fantom.id]: "fantom",
+};
+
+const useCheckApproval = (
```

Comment:
i wonder if we can handle approval check and approval without relying onto portals api.
I'm guessing it's possible if we first fetch the quote. Now we know the spender addy.
And we can handle approve logic on our own.
When approval not needed, we proceed with quote transaction.
I would check if it's possible here.

So the logic would be:
- useQuery get quote;
- usePortalsApproval - we handle ourself based on quote spender
- usePortalsSend - we handle ourself based on quote tx data, etc..

This way all portals stuff happens inside quote query. and we have mutations decoupled from queries